Story/cite 139

MaulikLim opened the pull request:
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed) 

[CITE-139] - Back and next/prev button on item page should work for search
  
## Anything else the reviewer needs to know?

... describe here ...


[CITE-139]: https://diging.atlassian.net/browse/CITE-139

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

NarmadaRavali reviewed (approved):
looks good to me

jdamerow commented:
While this was a probably also not working and needed fixing, this ticket referred to the next/prev buttons on the item page (not item**s** page). I don't think this will fix the issue.

MaulikLim commented:
Removed the previous API used for fetching prev and next citation item. Instead, fetched those details on item display itself. If the item is opened through search result, getting the next and previous items based on the search term.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ItemController.java:
```diff
@@ -25,34 +26,42 @@
         model.addAttribute("zoteroGroupId", zoteroGroupId);
         
+        CitationPage citationPage = null;
+        if (searchTerm.isEmpty() || searchTerm.equals(" ")) {
```
why not trim the searchterm and then check if it's empty?

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -13,7 +13,15 @@
+	            }
+	        } else {
+	            window.location.href = "items?page=" + page;
+	        }
```
I would take all query parameter and add them to the url, otherwise if we extend the functionality of the page and add more query params, we will have to change this again.

jdamerow replied:
?

MaulikLim replied:
I am not sure if any other parameter is applicable over here. I have added sort by. Only others are "collectionId" and "index" which does not seem to apply.

jdamerow replied:
yes, but in the future we might add functionality that adds more parameters

jdamerow replied:
what do you mean by "both of them"?

MaulikLim replied:
collectionId and index

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Let's also fix the back button since we're add it (to go back to the search results).

MaulikLim commented:
Addressed review comments. Fixed back button for item opened through search result. Also fixed it for an item opened from a collection.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -31,34 +31,31 @@
+				uri = next + "?index=" + [(${adjacentCitations.nextIndex})] + "&page=" + [(${adjacentCitations.nextPage})] + "&sortBy=[(${sortBy})]&collectionId=[(${collectionId})]";
+				if ("[(${searchTerm})]") {
+					uri = uri + "&searchTerm=[(${searchTerm})]";
 				}
```
Is there a reason this is not in an href attribute but handled in javascript?

MaulikLim replied:
No particular reason. I'll change this.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -25,6 +41,9 @@
+		else if (currentURL.match(/search/)) {
+			window.location.href = "items/" + key + "?searchTerm=[(${searchTerm})]" + "&index=" + index + "&page=[(${currentPage})]&sortBy=[(${sort})]&collectionId=[(${collectionId})]";
+		}
 		else{
```
let's make all these javascript if/else statements thymeleaf conditions; otherwise if we change the url or something like that, this javascript will stop working. It will also create less complex rendered html pages; and then use the code to get all the query parameters for all links (here and above for collection, search and other).

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ItemController.java:
```diff
@@ -25,34 +26,43 @@
+                        zoteroGroupId, collectionId, page, sortBy, Integer.valueOf(index));
+            }
+        } else {
+            citationPage = engine.getPrevAndNextCitation(searchTerm, zoteroGroupId, page - 1, Integer.valueOf(index), 50);
```
if you check for null if searchterm is empty, you would have to check here as well, dont' you? Shoudn't matter though once index has a default value.

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -13,23 +13,26 @@
+	                const params = new URLSearchParams(queryParams);
+	                params.delete('page');
+	                queryParams = "?" + params.toString();
+	            }
```
the above code should come first, then the check for the url (so all query parameters are always passed on)

MaulikLim replied:
I'm making this change. But now that I'm testing this, I find many places where these queryparams get lost. For e.g., when navigating between next and prev citations, or going back from item to items list, etc.

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -99,6 +102,20 @@
+            queryParams = params.toString();
+        }
+        
+        const x = document.getElementsByClassName("bib-entry");
```
is there a reason, you don't use jquery here?

jdamerow replied:
also, bad variable name, and why do you use `const` here?

MaulikLim replied:
No particular reason for not using jquery. I'll change it to jQuery.

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -99,6 +102,20 @@
+        }
+        
+        const x = document.getElementsByClassName("bib-entry");
+        for (const element of x) {
```
why does `element` need to be `const`?

MaulikLim commented:
Addresses review comments - Fixed mistake, replaced javascript with jquery for query param forwarding. Enabled query param forwarding when navigating between back, prev and next for an item page.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ItemController.java:
```diff
@@ -25,34 +26,41 @@
             List<String> fields = new ArrayList<>();
             citationManager.getItemTypeFields((IUser)authentication.getPrincipal(), citation.getItemType()).forEach(f -> fields.add(f.getFilename()));
             model.addAttribute("fields", fields);
+            if (citationPage == null) {
```
can `citationPage` ever be null? Does `citationManager.getPrevAndNextCitation` ever return null?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ItemController.java:
```diff
@@ -25,34 +26,41 @@
+            if (citationPage == null) {
+                citationPage = new CitationPage();
+            }
+            model.addAttribute("adjacentCitations", citationPage);
```
if `citationPage` can indeed be null, then merge this line with lines 54-56 by using the ternary operator.

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -27,38 +27,24 @@
+	$(document).ready(function() {
+	    var queryParams = window.location.search;
+	    if(queryParams)  {
+	        const params = new URLSearchParams(queryParams);
```
why does `params` have to be a const?

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -27,38 +27,24 @@
+	    $('.btn-back').attr("href", $('.btn-back').attr("href") + "&" + queryParams);
+	    $('.btn-prev').attr("href", $('.btn-prev').attr("href") + "&" + queryParams);
+	    $('.btn-next').attr("href", $('.btn-next').attr("href") + "&" + queryParams);
+	    console.log(queryParams);
```
remove

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -27,38 +27,24 @@
+	    
+	    $('.btn-back').attr("href", $('.btn-back').attr("href") + "&" + queryParams);
+	    $('.btn-prev').attr("href", $('.btn-prev').attr("href") + "&" + queryParams);
+	    $('.btn-next').attr("href", $('.btn-next').attr("href") + "&" + queryParams);
```
since these three lines are pretty much the same, just give them all the same class and have only one line with the class selelctor.

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -13,23 +13,28 @@
+	        if(!queryParams) {
+	            queryParams = "?";
+    	        } else {
+    	            const params = new URLSearchParams(queryParams);
```
why is this a const?

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -99,6 +104,21 @@
+$(document).ready(function() {
+    var queryParams = window.location.search;
+    if(queryParams)  {
+        const params = new URLSearchParams(queryParams);
```
why `const`?

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -99,6 +104,21 @@
+    if(queryParams)  {
+        const params = new URLSearchParams(queryParams);
+        params.delete('page');
+        params.delete('searchTerm');
```
is there a reason these are deleted? if they exist, they need to be added to all links, dont' they? and the same goes for the above delete statements.

MaulikLim replied:
We provide the new page number and also add the search term. They keep duplicating and probably have different values if we don't remove them.

jdamerow replied:
the searchterm would stay the same though, wouldn't it?

MaulikLim replied:
The user could add new search term and search again. All the relevant parameters like page, collectionId, search term, etc. are added wherever necessary. We need to remove them from previous URI to avoid any ambiguity and duplication.

jdamerow replied:
but that's only for the search input field, isn't it? For everything else, the searchTerm would be the same? Just make sure to remove the searchterm when you build the search url.

MaulikLim replied:
Previously when we didn't pass the parameters ahead, we already had the logic in place to add them to the URI with new values if necessary. This includes the searchTerm as well. So, the new URI will have those params included.

diging-jenkins commented:
Congratulations, build succeeded!

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
the `sortBy` parameter gets appended every time I click on a citation from the list view
`searchTerm=&index=30&page=2&sortBy=title,title,title,title&collectionId=&`

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!
